fix(curriculum): Added introduction to meta tag (#47221)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/14-responsive-web-design-22/learn-html-by-building-a-cat-photo-app/62bb4009e3458a128ff57d5d.md

---

> # --description--

Reviewer: I don't know how you bypass enlist errors, but this should have given you an error that you are missing an extra space 

```suggestion
# --description--

```

> To communicate meta data to the browser, a `<meta>` tag is used. An example of metadata is `charset`, which describes the character encoding of the document.

Reviewer: ```suggestion
You can set the browser behavior by using `meta` element. A famous use case is telling the browser to parse the markdown into multiple languages by setting `charset` attribute to `UTF-8`.
```

Author: sounds good. I've committed.

> In this step, a `meta` element will be added to the `head`. This self-closing element communicates metadata to the browser and can be written as `<meta>`. In this step, you will add a `charset` attribute which describes the character encoding to the browser.

Reviewer 2: I don't think we need to tell them that we are adding a meta element to the head in the first sentence because we are going to tell them to do that in the actual instructions. Same with telling them that they are going to add a charset attribute just before the instructions tell them to add a charset attribute.

# --description--

You can set browser behavior by adding self-closing `meta` elements in the `head`. Here's an example:

> ```

Reviewer: add `html` language

Author: I've updated the PR

> <meta attribute="value" />

Reviewer: Remove `/` as mentioned in the discussion

Reviewer 3: ```suggestion
<meta attribute="value">
```

# --description--

> You can set the browser behavior by using `meta` element. A famous use case is telling the browser to parse the markdown into multiple languages by setting `charset` attribute to `UTF-8`.

Reviewer 4: ````suggestion
You can set browser behavior by adding self-closing `meta` elements in the `head`. Here's an example:

```
<meta attribute="value" />
```
````

Reviewer 2: > and if they still use the closing tag, then just let it pass.

I'm not saying this would be the end of the world, but I don't think the HTML specs allow for a closing `</meta>` tag and the W3 HTML validator throws an error if you have one. My understanding is that we want to allow valid HTML wherever possible. Conversely, I think we should fail invalid HTML as well.

Reviewer 3: We've had a lot of back and forth on the `/>` for self-closing elements. Pretty sure we don't need the `/` and shouldn't include in the example?

__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :question: I have a question

Reviewer 4: > Pretty sure we don't need the / and shouldn't include in the example?

I'm fine with either method here.

> Create a `meta` element as a child of the `head` element. Set its `charset` attribute to `UTF-8`.

Reviewer 4: ```suggestion
Tell the browser to parse the markdown into multiple languages by creating a `meta` element as a child of the `head` element. Set its `charset` attribute to `UTF-8`.
```

# --hints--

You should create a self-closing `meta` element within the `head` element.

```js
assert.exists(document.querySelector('head > meta'));
let innerText = document.querySelector(" head ").innerHTML;
let matched = (innerText.match(/meta/g) || []).length

> assert.equal(matched, 1)

Reviewer: Hi @hanswang123456, the current test doesn't stop `<meta></meta>`

If you need help solving this, join us in [our chat room](https://discord.gg/PRyKn3Vbay) or [the forum](https://forum.freecodecamp.org/c/contributors/3)

Author: Sounds good, Ill join the chatroom

Reviewer 5: @hanswang123456 are you still stuck?

Author: Sorry for the late response, a challenge I've been running into is detecting the closing meta tag's existence. When users include the tag, it is ignored by the browser. I'll need to use another method other than innerHTML(which does not return the closing meta tag)

Reviewer 5: No worries - keep up the excellent work.

Reviewer 4: From other review thread:

> I don't think the HTML specs allow for a closing </meta> tag and the W3 HTML validator throws an error if you have one. My understanding is that we want to allow valid HTML wherever possible. Conversely, I think we should fail invalid HTML as well.

While the validator may fail with this, the browser still renders the original meta tag as expected and just ignores the closing tag. Typically, I might say that we don't need to test things like this because the browser gives us what we expect and we can't test every edge case. But this is the first project, I would be fine with a test for it. 

I would add a third test for it @hanswang123456, the first test can be reverted to what it was (just check that the meta tag exists). The second one can stay as it is. And the third test can be `You should not have a closing meta tag` or `You should not have a closing </meta> tag`. The test will have to just use regex since nothing is rendered if a closing tag is added.

Author: > From other review thread:
> 
> > I don't think the HTML specs allow for a closing  tag and the W3 HTML validator throws an error if you have one. My understanding is that we want to allow valid HTML wherever possible. Conversely, I think we should fail invalid HTML as well.
> 
> While the validator may fail with this, the browser still renders the original meta tag as expected and just ignores the closing tag. Typically, I might say that we don't need to test things like this because the browser gives us what we expect and we can't test every edge case. But this is the first project, I would be fine with a test for it.
> 
> I would add a third test for it @hanswang123456, the first test can be reverted to what it was (just check that the meta tag exists). The second one can stay as it is. And the third test can be `You should not have a closing meta tag` or `You should not have a closing </meta> tag`. The test will have to just use regex since nothing is rendered if a closing tag is added.

Sounds good. I'll update the 3rd test .

Reviewer: Thank you @hanswang123456, I have update the test to close the issue. Thank you for the hard work and bearing with us

assert.equal(document.querySelector('head > meta')?.getAttribute('charset')?.toLowerCase(), 'utf-8');
```

Your `meta` element should be a self-closing tag, you don't need to add `</meta>`.

```js

> assert(code.match(/<\/meta\s*>?/i) === null);

Reviewer 3: ```suggestion
assert.notMatch(code, /<\/meta\s*>?/i);
```

We can use the specific assert API method for this.
